src/core/qgsvectorlayerutils.cpp

```diff
@@ -380,8 +380,16 @@ QgsFeature QgsVectorLayerUtils::createFeature( const QgsVectorLayer *layer, cons
     bool checkUnique = true;
 
     // in order of priority:
+    // 1. passed attribute value
+    if ( attributes.contains( idx )
+         && !layer->primaryKeyAttributes().contains( idx )
+         && !( fields.at( idx ).constraints().constraints() & QgsFieldConstraints::ConstraintUnique ) )
+    {
+      v = attributes.value( idx );
+    }
 
-    // 1. client side default expression
+    // 2. client side default expression
+    // note - deliberately not using else if!
     if ( layer->defaultValueDefinition( idx ).isValid() )
```

Comment:
this needs to be adjusted to if `( !v.isValid() && layer-> .... )` or it will overwrite any manually set attributes.

Reply:
Done on commit 9a67fc23ba334a2dca8e78a320a095caac2bea76